R/knitr-cache.R

```diff
@@ -0,0 +1,101 @@
+#' A reticulate cache engine for Knitr
+#'
+#' This provides a `reticulate` cache engine for `knitr`. The cache engine
+#' allows `knitr` to save and load Python sessions between cached chunks. The
+#' cache engine depends on the `dill` Python module. Therefore, you must have
+#' `dill` installed in your Python environment.
+#'
+#' The engine can be activated by setting (for example)
+#'
+#' ```
+#' knitr::cache_engines$set(python = reticulate::cache_eng_python)
+#' ```
+#'
+#' Typically, this will be set within a document's setup chunk, or by the
+#' environment requesting that Python chunks be processed by this engine.
+#'
+#' @param options
+#'   List of chunk options provided by `knitr` during chunk execution.
+#'   Contains the caching path.
+#'
+#' @export
+cache_eng_python <- (function() {
+  check_cache_available <- function(options) {
+    eng_python_initialize(options)
+
+    # does the python version is supported by 'dill'?
+    if (py_version() < "3.7") {
+      warning("Python cache requires Python version >= 3.7")
+      return(FALSE)
+    }
+
+    # is the module 'dill' loadable?
+    dill <- tryCatch(import("dill"), error = identity)
+    if (inherits(dill, "error")) {
+      error <- reticulate::py_last_error()
+      if (!error$type %in% c("ImportError", "ModuleNotFoundError"))
+        stop(error$value, call. = FALSE)
+      warning("The Python module 'dill' was not found, it's required for Python cache")
```

Comment:
Can we use the same warning re: requiring dill >= 0.3.6?

Reply:
Yup! I'll try to simplify the code.

Reply:
Looks better now.